File: 2074-ReverseNodesinEvenLengthGroups/2074-ReverseNodesinEvenLengthGroups.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseEvenLengthGroups = function(head) {
    let curr = head;
    let dummy = new ListNode();
    dummy.next = head;
    let prev = dummy;
    groupSize = 1;

    while(curr){
        actualGroupSize = 0;
        temp = curr;
        for(let i = 0; i < groupSize; i++){
            if(temp){
                temp = temp.next;
                actualGroupSize++;
            } else{
                break;
            }
        }

        let startGroup = curr;
        let endGroup = curr;

        for(let i = 0; i < (actualGroupSize - 1); i++){
            endGroup = endGroup.next;
        }

        if(actualGroupSize % 2 == 0){
            let prevGroup = prev;
            prev = null;
            let tailGroup = endGroup.next;

            for(let i = 0; i < actualGroupSize; i++){
                let next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
            }

            prevGroup.next = prev;
            startGroup.next = tailGroup;
            prev = startGroup;
            
        } else {
            for(let i = 0; i < actualGroupSize; i++){
                prev = curr;
                curr = curr.next;
            }
        }
        groupSize++;
    }
    
    return dummy.next;
};

File: 2074-ReverseNodesinEvenLengthGroups/2074-ReverseNodesinEvenLengthGroups.py
class Solution:
    def reverseEvenLengthGroups(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Initialise group size to 1
        group_size = 1

        # Initialize dummy node and connect it
        dummy = ListNode(0)
        dummy.next = head

        # Initialize previous and current pointers
        prev = dummy
        curr = head

        while curr:
            # Calculate the actual group size
            temp = curr
            actual_group_size = 0
            for _ in range(group_size):
                if temp:
                    temp = temp.next
                    actual_group_size += 1
                else:
                    break
            
            # Track the start of the group
            start_group = curr
            end_group = curr
            for _ in range(actual_group_size - 1):
                end_group = end_group.next

            if actual_group_size % 2 == 0:
                # Reverse the group
                prev_group = prev
                prev = None
                tail_group = end_group.next  # The node after the group

                for _ in range(actual_group_size):
                    next_node = curr.next
                    curr.next = prev
                    prev = curr
                    curr = next_node
                
                # Connect the reversed group to the rest of the list
                prev_group.next = prev
                start_group.next = tail_group
                prev = start_group # **Fix: make sure the tail is connected properly**

            else:
                # Move the pointers for odd-length groups
                for _ in range(actual_group_size):
                    prev = curr
                    curr = curr.next

            # Increment group size for the next iteration
            group_size += 1   

        return dummy.next
